src/combinatorics/simple_graph/metric.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Vincent Beffara
+-/
+import combinatorics.simple_graph.connectivity
+import data.nat.lattice
+
+/-!
+# Graph metric
+
+This module defines the `simple_graph.dist` function, which takes
+pairs of vertices to the length of the shortest walk between them.
+
+## Main definitions
+
+- `simple_graph.dist` is the graph metric.
+
+## Tags
+
+graph metric
+
+-/
+
+namespace simple_graph
+variables {V : Type*} (G : simple_graph V)
+
+/-! ## Metric -/
+
+/-- The distance between two vertices is the length of the shortest walk between them.
+If no such walk exists, this uses the junk value of `0`. -/
+noncomputable
+def dist (u v : V) : ℕ := Inf (set.range (walk.length : G.walk u v → ℕ))
+
+variables {G}
+
+lemma reachable.range_length_walk_nonempty {u v : V} (hr : G.reachable u v) :
+  (set.range (walk.length : G.walk u v → ℕ)).nonempty :=
+begin
```

Comment:
What is wrong with
```
lemma reachable.range_length_walk_nonempty {u v : V} (hr : G.reachable u v) :
  (set.range (walk.length : G.walk u v → ℕ)).nonempty :=
set.range_nonempty_iff_nonempty.mpr hr
```

Reply:
Are you asking "this works, why didn't you do it?" or "I thought this would work, but it doesn't for some reason?" or something else?

It's hard to answer the first question without knowing why you think what you wrote is better. There's usually nothing wrong with two-line proofs, but if a reviewer kindly supplies a golfed version, then so long as it works and doesn't violate API, then great. I'll assume you're suggesting a golfing opportunity here and test it out.

Reply:
Thanks, I missed `set.range_nonempty_iff_nonempty`

Reply:
I should have phrased my comment better, sorry. The initial reason for my question was that I noticed `reachable_iff_nonempty_univ` in the other file and was wondering where it was used. And the reason for wondering that is that I am still uncomfortable with coercions and casts and all that, and in my mental model `univ` is in the same place as these.

Reply:
Lemmas like `reachable_iff_nonempty_univ` are there to flesh out the definition of `reachable`, to help you not have to unfold the definition before rewriting. I had it go into the language of sets because manipulating instances (like `nonempty`) can be a pain due to the fact Lean freezes the instance cache by default.

These `range_length_walk_nonempty` lemmas seem like their actually not so useful, so I just folded the one instance of it into where it was used.

Don't worry, there are no coercions involved with `set.univ` here.  Maybe the type ascription is making you think about it that way? If you just write `set.univ.nonempty`, then Lean can't figure out what to use for the implicit argument, and `(set.univ : set X).nonempty` is a way around that.  Lean inserts casts at type ascriptions if there's a typechecking error there, but otherwise they're just hints.  The existence of both `nonempty` and `set.nonempty` I guess are a way to avoid casts one way or the other.